gpu4pyscf/dft/numint.py

```diff
@@ -1523,36 +1679,41 @@ def _block_loop(ni, mol, grids, nao=None, deriv=0, max_memory=2000,
     if opt is None or mol not in [opt.mol, opt._sorted_mol]:
         ni.build(mol, grids.coords)
         opt = ni.gdftopt
-
+    
     mol = None
     _sorted_mol = opt._sorted_mol
+    if not hasattr(grids, 'sparse_cache') or len(grids.sparse_cache) == 0:
+        grids.build_sparsity(_sorted_mol)
+    
+    ngrids = grids.coords.shape[0]
+    ao_indices, s_index, ao_loc_non0, nao_non0 = grids.sparse_cache[blksize, ngrids]
     with opt.gdft_envs_cache():
-        block_id = 0
         t1 = log.init_timer()
-        for ip0, ip1 in lib.prange(0, ngrids, blksize):
+        for block_id, (ip0, ip1) in enumerate(lib.prange(0, ngrids, blksize)):
             coords = grids.coords[ip0:ip1]
             weight = grids.weights[ip0:ip1]
             t1 = log.init_timer()
-            # cache ao indices
-            if (block_id, blksize, ngrids) not in ni.non0ao_idx:
-                ni.non0ao_idx[block_id, blksize, ngrids] = _sparse_index(_sorted_mol, coords, opt.l_ctr_offsets)
+            #pad, idx, non0tab, ao_loc_non0 = grids.sparse_cache[block_id, blksize, ngrids]
+            #ao_loc_non0 = grids.ao_loc_non0[block_id]
+            #idx = grids.ao_indices[block_id]
+            #non0tab = grids.screen_index[block_id]
+            nao_pad = (nao_non0[block_id] + AO_ALIGNMENT - 1) // AO_ALIGNMENT * AO_ALIGNMENT
+            nao_pad = min(nao_pad, nao)
+            pad = nao_pad - nao_non0[block_id]
 
-            pad, idx, non0shl_idx, ctr_offsets_slice, ao_loc_slice = ni.non0ao_idx[block_id, blksize, ngrids]
             ao_mask = eval_ao(
                 ni, _sorted_mol, coords, deriv,
-                nao_slice=len(idx),
-                shls_slice=non0shl_idx,
-                ao_loc_slice=ao_loc_slice,
-                ctr_offsets_slice=ctr_offsets_slice)
-
+                nao_non0=nao_pad,
+                ao_loc_non0=ao_loc_non0[block_id],
+                non0tab=s_index[block_id])
             t1 = log.timer_debug2('evaluate ao slice', *t1)
             if pad > 0:
                 if deriv == 0:
                     ao_mask[-pad:,:] = 0.0
                 else:
                     ao_mask[:,-pad:,:] = 0.0
-            block_id += 1
-            yield ao_mask, idx, weight, coords
+            idx = ao_indices[block_id][:nao_pad]
+            yield ao_mask, idx, weight, nao_pad
```

Comment:
Do not change block_loop api. Rather, create a new "???_loop" function that returns the sparsity table and all required information.

Reply:
The PR is not ready for review yet.

Reply:
There is a function called `_batch_loop`. But the changes are still messy for reviewing.